File: pwiz/analysis/demux/MSXDemultiplexer.cpp
//
// $Id$
//
//
// Original author: Jarrett Egertson <jegertso .@. uw.edu>
//
// Licensed under the Apache License, Version 2.0 (the "License"); 
// you may not use this file except in compliance with the License. 
// You may obtain a copy of the License at 
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software 
// distributed under the License is distributed on an "AS IS" BASIS, 
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
// See the License for the specific language governing permissions and 
// limitations under the License.
//

#include "MSXDemultiplexer.hpp"
#include "SpectrumPeakExtractor.hpp"
#include "DemuxHelpers.hpp"


namespace pwiz {
namespace analysis {
    using namespace std;
    using namespace DemuxTypes;
    using namespace msdata;

    MSXDemultiplexer::MSXDemultiplexer(Params p) :
        params_(p)
    {
    }

    MSXDemultiplexer::~MSXDemultiplexer()
    {
    }

    void MSXDemultiplexer::Initialize(SpectrumList_const_ptr sl, IPrecursorMaskCodec::const_ptr pmc)
    {
        assert(sl);
        assert(pmc);
        if (!sl || !pmc)
            return;
        sl_ = sl;
        pmc_ = pmc;
    }

    void MSXDemultiplexer::BuildDeconvBlock(size_t index, const vector<size_t>& muxIndices, MatrixPtr& masks, MatrixPtr& signal) const
    {
        assert(sl_);
        assert(pmc_);
        if (!sl_ || !pmc_)
            throw runtime_error("Null pointer to SpectrumList and/or IPrecursorMaskCodec, MSXDemultiplexer may not have been initialized.");

        // get the list of peaks to demultiplex
        BinaryDataArrayPtr mzsToDemux = sl_->spectrum(index, true)->getMZArray();
        SpectrumPeakExtractor peakExtractor(mzsToDemux->data, params_.massError);

        // initialize mask and intensities matrices
        masks.reset(new MatrixType(muxIndices.size(), pmc_->GetDemuxBlockSize()));
        signal.reset(new MatrixType(muxIndices.size(), mzsToDemux->data.size()));
        
        int specPerCycle = pmc_->GetSpectraPerCycle();
        for (int matrixRow = 0; matrixRow < muxIndices.size(); ++matrixRow)
        {
            size_t currentIndex = muxIndices[matrixRow];
            Spectrum_const_ptr s = sl_->spectrum(currentIndex, true);
            DemuxScalar weight = 1.0;
            if (params_.applyWeighting)
            {
                /* This method of weighting tries to minimize the added variance from changing intensity during
                * chromatogram elution. Given a point on the elution peak, one can model the change in intensity as a
                * function of difference in retention time. By modeling this function for all possible points on the
                * peak it is possible to find a best fit weighting scheme that minimizes intensity variance. To
                * describe the model itself in more detail, ScanDiff contains the difference in retention time and
                * 5 / specPerCycle contains information about the typical peak width (or standard deviation).
                */
                int scanDiff = int(index) - int(currentIndex);
                weight = 1.0 / (1.0 + pow(5.0 * scanDiff / specPerCycle, 2));
            }
            pmc_->GetMask(s, *masks, matrixRow, weight);
            if (params_.variableFill)
            {
                // the intensity values in a variable fill scan are written out
                // as the # ions / total injection time (summed)
                double totalInjectionTime = 0.0;
                for (const auto& p : s->precursors)
                {
                    auto injectParam = p.userParam("MultiFillTime");
                    if (injectParam.empty())
                    {
                        throw runtime_error("[SpectrumlList_MsxDemux] Tried to process an MS2 scan without multi-fill times written using variable fill demux ");
                    }
                    totalInjectionTime += injectParam.valueAs<double>();
                }
                weight *= totalInjectionTime / 1000.0;    // need total injection time in seconds
            }
            peakExtractor(s, *signal, matrixRow, weight);
        }

        // cache the spectrum indices
        Spectrum_const_ptr sPtr = sl_->spectrum(index, true);
        pmc_->SpectrumToIndices(sPtr, spectrumIndices_);
    }

    void MSXDemultiplexer::GetMatrixBlockIndices(size_t indexToDemux, std::vector<size_t>& muxIndices, double demuxBlockExtra) const
    {
        demuxBlockExtra = max(0.0, demuxBlockExtra);
        auto numSpectraToFind = pmc_->GetDemuxBlockSize() + size_t(round(demuxBlockExtra * pmc_->GetSpectraPerCycle()));
        if (!FindNearbySpectra(muxIndices, sl_, indexToDemux, numSpectraToFind))
            throw runtime_error("GetMatrixBlockIndices() Not enough spectra to demultiplex this block");
    }

    const std::vector<size_t>& MSXDemultiplexer::SpectrumIndices() const
    {
        return spectrumIndices_;
    }
} // namespace analysis
} // namespace pwiz

File: pwiz-sharp/pwiz/src/Analysis/Demux/MsxDemultiplexer.cs
using MathNet.Numerics.LinearAlgebra;
using MathNet.Numerics.LinearAlgebra.Double;
using Pwiz.Data.Common.Cv;
using Pwiz.Data.MsData.Spectra;
using Pwiz.Util.Chemistry;

namespace Pwiz.Analysis.Demux;

/// <summary>
/// MSX-style demultiplexer. Port of cpp's <c>MSXDemultiplexer</c>. Use this for MSX
/// experiments (with or without overlap); for pure overlap data prefer
/// <see cref="OverlapDemultiplexer"/> which adds chromatographic interpolation.
/// </summary>
public sealed class MsxDemultiplexer : IDemultiplexer
{
    /// <summary>Tunable parameters.</summary>
    public sealed class Params
    {
        /// <summary>Mass tolerance for MS/MS peak extraction (cpp default 10 ppm).</summary>
        public MZTolerance MassError { get; init; } = new(10.0, MZToleranceUnits.Ppm);

        /// <summary>Down-weight nearby spectra by their retention-time distance from the spectrum
        /// being demuxed. Cpp default true.</summary>
        public bool ApplyWeighting { get; init; } = true;

        /// <summary>Set when fill times vary per scan window — multiplies each row's weight by
        /// the precursor's <c>MultiFillTime</c> sum / 1000 (seconds).</summary>
        public bool VariableFill { get; init; }
    }

    private readonly Params _params;
    private ISpectrumList? _spectrumList;
    private IPrecursorMaskCodec? _maskCodec;
    private List<int> _spectrumIndices = new();

    /// <summary>Constructs an MSX demultiplexer with the given parameters.</summary>
    public MsxDemultiplexer(Params? p = null)
    {
        _params = p ?? new Params();
    }

    /// <inheritdoc/>
    public IReadOnlyList<int> SpectrumIndices => _spectrumIndices;

    /// <inheritdoc/>
    public void Initialize(ISpectrumList spectrumList, IPrecursorMaskCodec maskCodec)
    {
        ArgumentNullException.ThrowIfNull(spectrumList);
        ArgumentNullException.ThrowIfNull(maskCodec);
        _spectrumList = spectrumList;
        _maskCodec = maskCodec;
    }

    /// <inheritdoc/>
    public void BuildDeconvBlock(int spectrumIndex, IReadOnlyList<int> muxIndices,
        out Matrix<double> masks, out Matrix<double> signal)
    {
        if (_spectrumList is null || _maskCodec is null)
            throw new InvalidOperationException(
                "Null SpectrumList or IPrecursorMaskCodec; MsxDemultiplexer may not have been initialized.");

        var deconvSpectrum = _spectrumList.GetSpectrum(spectrumIndex, getBinaryData: true);
        var mzArr = deconvSpectrum.GetMZArray() ?? throw new InvalidOperationException("Deconv spectrum has no m/z array");
        var peakExtractor = new SpectrumPeakExtractor(mzArr.Data, _params.MassError);

        masks = DenseMatrix.Create(muxIndices.Count, _maskCodec.DemuxBlockSize, 0);
        signal = DenseMatrix.Create(muxIndices.Count, mzArr.Data.Count, 0);

        int specPerCycle = _maskCodec.SpectraPerCycle;
        for (int row = 0; row < muxIndices.Count; row++)
        {
            int currentIndex = muxIndices[row];
            var s = _spectrumList.GetSpectrum(currentIndex, getBinaryData: true);
            double weight = 1.0;
            if (_params.ApplyWeighting)
            {
                // cpp: weight = 1 / (1 + (5 * scanDiff / specPerCycle)^2). Damps far-away spectra
                // assuming a roughly Gaussian elution-peak shape with σ ≈ specPerCycle/5.
                int scanDiff = spectrumIndex - currentIndex;
                double t = 5.0 * scanDiff / specPerCycle;
                weight = 1.0 / (1.0 + t * t);
            }
            _maskCodec.GetMask(s, masks, row, weight);
            if (_params.VariableFill)
            {
                double totalInjectionTime = 0;
                foreach (var p in s.Precursors)
                {
                    var injectParam = p.UserParam("MultiFillTime");
                    if (injectParam.IsEmpty)
                        throw new InvalidOperationException(
                            "[MsxDemultiplexer] MS2 scan missing MultiFillTime user param required by variableFill demux.");
                    totalInjectionTime += injectParam.ValueAs<double>();
                }
                weight *= totalInjectionTime / 1000.0;
            }
            peakExtractor.Extract(s, signal, row, weight);
        }

        _spectrumIndices = new List<int>();
        _maskCodec.SpectrumToIndices(deconvSpectrum, _spectrumIndices);
    }

    /// <inheritdoc/>
    public void GetMatrixBlockIndices(int indexToDemux, List<int> muxIndices, double demuxBlockExtra = 0.0)
    {
        if (_spectrumList is null || _maskCodec is null)
            throw new InvalidOperationException(
                "Null SpectrumList or IPrecursorMaskCodec; MsxDemultiplexer may not have been initialized.");

        demuxBlockExtra = System.Math.Max(0.0, demuxBlockExtra);
        int numSpectraToFind = _maskCodec.DemuxBlockSize +
            (int)System.Math.Round(demuxBlockExtra * _maskCodec.SpectraPerCycle);
        if (!DemuxHelpers.FindNearbySpectra(muxIndices, _spectrumList, indexToDemux, numSpectraToFind))
            throw new InvalidOperationException(
                "GetMatrixBlockIndices() Not enough spectra to demultiplex this block");
    }
}
